driver-core/src/main/com/mongodb/internal/operation/ClientBulkWriteOperation.java

```diff
@@ -0,0 +1,1051 @@
+            final boolean effectiveRetryWrites,
+            final WriteConcern effectiveWriteConcern,
+            final SessionContext sessionContext,
+            final List<? extends ClientNamespacedWriteModel> unexecutedModels,
+            final BatchEncoder batchEncoder,
+            final Runnable ifCommandIsRetryable) {
+        // BULK-TODO This implementation must limit the number of `models` it includes in a batch if needed.
+        // Each batch re-selects a server and re-checks out a connection because this is simpler and it is allowed,
+        // see https://mongodb.slack.com/archives/C035ZJL6CQN/p1722265720037099?thread_ts=1722264610.664109&cid=C035ZJL6CQN.
+        return new BsonDocumentWrapper<>(
+                BULK_WRITE_COMMAND_NAME,
+                new Encoder<String>() {
+                    @Override
+                    public void encode(final BsonWriter writer, final String commandName, final EncoderContext encoderContext) {
+                        batchEncoder.reset();
+                        writer.writeStartDocument();
+                        writer.writeInt32(commandName, 1);
+                        writer.writeBoolean("errorsOnly", !options.isVerboseResults());
+                        writer.writeBoolean("ordered", options.isOrdered());
+                        options.isBypassDocumentValidation().ifPresent(value -> writer.writeBoolean("bypassDocumentValidation", value));
+                        options.getComment().ifPresent(value -> {
+                            writer.writeName("comment");
+                            encodeUsingRegistry(writer, value);
+                        });
+                        options.getLet().ifPresent(value -> {
+                            writer.writeName("let");
+                            encodeUsingRegistry(writer, value);
+                        });
+                        Function<ClientWriteModel, Boolean> modelSupportsRetries = model ->
+                                !(model instanceof ClientUpdateManyModel || model instanceof ClientDeleteManyModel);
+                        assertFalse(unexecutedModels.isEmpty());
+                        LinkedHashMap<MongoNamespace, Integer> indexedNamespaces = new LinkedHashMap<>();
+                        writer.writeStartArray("ops");
+                        boolean commandIsRetryable = effectiveRetryWrites;
+                        for (int modelIndexInBatch = 0; modelIndexInBatch < unexecutedModels.size(); modelIndexInBatch++) {
+                            ConcreteClientNamespacedWriteModel modelWithNamespace = getModelWithNamespace(unexecutedModels, modelIndexInBatch);
+                            ClientWriteModel model = modelWithNamespace.getModel();
+                            if (commandIsRetryable && !modelSupportsRetries.apply(model)) {
+                                commandIsRetryable = false;
+                                logWriteModelDoesNotSupportRetries();
```

Comment:
We can. I think, that's how it is done.

Reply:
It seems like it will log once per model.  Am I misunderstanding the code?

Reply:
Here is how I understand the code:

1. When `!modelSupportsRetries.apply(model)` is true, i.e., `model` does not support retries, we enter the `if` block.
2. There we do `commandIsRetryable = false` and log.
3. The `if` condition can never be `true` again for the same batch, because it starts with `commandIsRetryable && ...`.
4. Thus, we log at most once per batch.

Reply:
Ah, ok